### PR TITLE
network: introduce UplinkInterface= in [IPv6SendRA]

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

**Comment by yuwata** [comment]
Hm, CentOS CI (Arch Linux) failure is
```
testsuite-50.sh[389]: Failed to dissect image '/tmp/tmp.W8g91S/minimal_0.gpt': Operation not permitted
```

**Comment by mrc0mmand** [comment]
> Hm, CentOS CI (Arch Linux) failed with the following:
> 
> ```
> testsuite-50.sh[389]: Failed to dissect image '/tmp/tmp.W8g91S/minimal_0.gpt': Operation not permitted
> ```

*Thankfully*, the real issue was:

```
[   62.187564] testsuite-36.sh[219]: + sleep 1
[   62.190388] audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=numa-test comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
[   62.206126] systemd[529]: numa-test.service: Failed to set NUMA memory policy: Invalid argument
[   62.206304] systemd[529]: numa-test.service: Failed at step NUMA_POLICY spawning /bin/sleep: Invalid argument
[   62.206859] systemd[1]: Started numa-test.service.
[   62.208854] systemd[1]: numa-test.service: Failed with result 'exit-code'.
```

in TEST-36-NUMAPOLICY. Not sure what happened there, but I restarted the job, for now, since it's definitely unrelated.

File: src/network/networkd-queue.c [hunk 1]
```diff
@@ -118,6 +120,9 @@ static void request_hash_func(const Request *req, struct siphash *state) {
         case REQUEST_TYPE_NEXTHOP:
                 nexthop_hash_func(req->nexthop, state);
                 break;
+        case REQUEST_TYPE_RADV:
+                /* This type does not have object. */
```
Comment (keszybz) [inline src/network/networkd-queue.c]: "an object"

File: src/network/networkd-radv.c [hunk 2]
```diff
@@ -779,6 +771,69 @@ int radv_update_mac(Link *link) {
...
+        assert(link->network);
+
+        if (!IN_SET(link->state, LINK_STATE_CONFIGURING, LINK_STATE_CONFIGURED))
+                return false;
+
+        if (link->network->router_emit_dns && link->network->router_dns)
+                for (size_t i = 0; i < link->network->n_router_dns; i++)
+                        if (in6_addr_is_null(&link->network->router_dns[i]) &&
+                            in6_addr_is_null(&link->ipv6ll_address))
+                                return false;
```
Comment (keszybz) [inline src/network/networkd-radv.c]: This looks strange. Shouldn't the check for `in6_addr_is_null(&link->ipv6ll_address)` be moved out of the loop?

File: src/network/networkd-radv.c [hunk 3]
```diff
@@ -779,6 +771,69 @@ int radv_update_mac(Link *link) {
...
+
+        if (!link_radv_enabled(link))
+                return 0;
+
+        if (link->radv)
+                return 0;
+
+        r = link_queue_request(link, REQUEST_TYPE_RADV, NULL, false, NULL, NULL, NULL);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to request to configure IPv6 Router Advertisement engine: %m");
```
Comment (keszybz) [inline src/network/networkd-radv.c]: "Failed to request configuration of IPv6 Router Advertisement engine"

File: src/network/networkd-radv.c [hunk 4]
```diff
@@ -779,6 +771,69 @@ int radv_update_mac(Link *link) {
...
+
+        log_link_debug(link, "Requested to configure IPv6 Router Advertisement engine.");
```
Comment (keszybz) [inline src/network/networkd-radv.c]: "Requested configuration of …"

File: man/systemd.network.xml [hunk 5]
```diff
@@ -2584,30 +2584,41 @@ IPv6Token=prefixstable:2002:da8:1::</programlisting></para>
         for details. Defaults to <literal>medium</literal>.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>UplinkInterface=</varname></term>
+        <listitem><para>Specifies name or index of uplink interface, or one of the special values
```
Comment (keszybz) [inline man/systemd.network.xml]: "Specifies the name or the index of the uplink interface…"

File: man/systemd.network.xml [hunk 6]
```diff
@@ -2584,30 +2584,41 @@ IPv6Token=prefixstable:2002:da8:1::</programlisting></para>
...
+        <literal>:none</literal> and <literal>:auto</literal>. When emitting DNS servers or search
+        domains are enabled but no servers are specified, the servers configured in the uplink
```
Comment (keszybz) [inline man/systemd.network.xml]: are enabled → is enabled

File: man/systemd.network.xml [hunk 7]
```diff
@@ -2584,30 +2584,41 @@ IPv6Token=prefixstable:2002:da8:1::</programlisting></para>
...
+        interface will be emitted. When <literal>:auto</literal>, the link which has default gateway
+        with higher priority will be automatically selected. When <literal>:none</literal>, no uplink
```
Comment (keszybz) [inline man/systemd.network.xml]: "the link which has a default gateway with the highest priority will be automatically selected"

File: src/network/networkd-network.c [hunk 8]
```diff
@@ -1170,6 +1171,71 @@ int config_parse_link_group(
...
+                           "Invalid interface name in %s=, ignoring assignment: %s", lvalue, rvalue);
+                return 0;
+        }
+
+        /* The interface name will be resolved later. */
+        r = free_and_strdup_warn(name, rvalue);
+        if (r < 0)
+                return r;
+
+        *index = 0;
```
Comment (keszybz) [inline src/network/networkd-network.c]: Hmm, this is the same as `UPLINK_INDEX_AUTO`. Will this not cause confusion? I guess the code below first looks at `router_uplink_name`, so it'll be OK. But even in that case I think it's better to use `= UPLINK_INDEX_AUTO` here.

**Review by keszybz — COMMENTED** [review]
Looks OK. But let's merge this after the release. Maybe rebase?

File: src/network/networkd-radv.c (hunk 3 above)
Comment (yuwata) [reply src/network/networkd-radv.c]: Hmm, "Failed to request configuration" sounds also strange for me...

How about "Failed to enqueue a request to configure IPv6 Router Advertisement engine: %m"?

Comment (yuwata) [reply src/network/networkd-radv.c]: Or, "Failed to request IPv6 Router Advertisement engine: %m"?

File: src/network/networkd-radv.c (hunk 4 above)
Comment (yuwata) [reply src/network/networkd-radv.c]: How about "IPv6 Router Advertisement engine is requested."?

**Comment by yuwata** [comment]
@keszybz Thank you for your review. Updated. Most comments are addressed. Could you re-check the log messages you commented on?

File: src/network/networkd-radv.c (hunk 3 above)
Comment (keszybz) [reply src/network/networkd-radv.c]: Maybe "Failed to request configuring of IPv6 Router Advertisement engine", as discussed in the other PR.

**Comment by keszybz** [comment]
LGTM. In #19939, whatever final form for the error messages is decided, please just use the same here, if different from what is now merged.

**Pull request merged**